### PR TITLE
TriSolver (dist): move sorting permutation from CPU to GPU

### PR DESCRIPTION
**Pull request by albestro**
This PR aims at dropping the custom `permuteJustLocal` and reduce the use-case, by transforming permutation indices, to be manageable with the existing local permutation implementation, that exists for both backends.

- [x] Cleanup implementation
- [x] ~It might be possible to drop i5 (for distributed implementation)~
- [ ] What to do about `permute` API? Should we separate the "distributed" use case (at least formally) or is it enough reviewing assumptions?
- [x] ~Evaluate if it is worth switching to `MatrixRef` (just for the code changed)~
- [x] Make it work on GPU
- [x] Add a unit test for the new use-case with distributed matrices

### Notes
From PR #967 each rank sort eigenvalues by type (upper, dense, lower, deflated) independently from other ranks. At the time of that PR, for convenience reasons, we opted for performing the sort with a custom permutation procedure `permuteJustLocal` that were able to deal with global indices but just apply the permutation to the local part. In addition to this, `permuteJustLocal` was implemented just on CPU because on GPU it would had required a major effort not worth due to the inherently GPU inefficient type of operations.

### CODE REVIEW

**Comment by albestro** [comment]
cscs-ci run

**Comment by albestro** [comment]
cscs-ci run

**Comment by albestro** [comment]
cscs-ci run

**Comment by albestro** [comment]
cscs-ci run

**Comment by albestro** [comment]
cscs-ci run

**Comment by albestro** [comment]
cscs-ci stop

**Comment by albestro** [comment]
cscs-ci run

File: include/dlaf/eigensolver/tridiag_solver/merge.h [hunk 1]
```diff
@@ -394,23 +397,28 @@ auto stablePartitionIndexForDeflationArrays(const SizeType n, const ColType* typ
...
-// @param evals                 array[n] of eigenvalues sorted as perm_sorted
-// @param perm_sorted           array[n] current -> initial (i.e. evals[i] -> types[perm_sorted[i]])
-// @param index_sorted          array[n] global(sort(non-deflated)|sort(deflated))) -> initial
-// @param index_sorted_coltype  array[n] local(sort(upper)|sort(dense)|sort(lower)|sort(deflated))) -> initial
+// @param types                 array[n]    column type of each eigenvector after deflation (initial order)
+// @param evals                 array[n]    of eigenvalues sorted as perm_sorted
+// @param perm_sorted           array[n]    current -> initial (i.e. evals[i] -> types[perm_sorted[i]])
+// @param index_sorted          array[n]    global(sort(non-deflated)|sort(deflated))) -> initial
+// @param index_sorted_coltype  array[n]    local(sort(upper)|sort(dense)|sort(lower)|sort(deflated))) -> initial
+// @param i5_lc                 array[n_lc] local(sort(upper)|sort(dense)|sort(lower)|sort(deflated))) -> initial
```
Comment (albestro) [inline include/dlaf/eigensolver/tridiag_solver/merge.h:421]: note-to-self: specify that they are local indices, while in `index_sorted_coltype` they are global indices

**Comment by albestro** [comment]
cscs-ci run

**Review by rasolca — APPROVED** [review]
Looks good.

**Comment by rasolca** [comment]
cscs-ci run

File: include/dlaf/permutations/general/impl.h [hunk 2]
```diff
@@ -458,39 +381,21 @@ void applyPackingIndex(const matrix::Distribution& subm_dist, IndexMapSender&& i
...
-                                  mat_in_tiles, mat_out_tiles);
+      applyPermutationOnCPU<T, C>(i_perm, subm_dist, perm_arr, mat_in_tiles, mat_out_tiles);
     };
 
-    ex::start_detached(std::move(sender) |
-                       dlaf::internal::transform(dlaf::internal::Policy<Backend::MC>(),
-                                                 std::move(setup_permute_fn)) |
-                       ex::unpack() | ex::bulk(subm_dist.size().get<C>(), permute_fn));
+    ex::start_detached(std::move(sender) | ex::transfer(di::getBackendScheduler<Backend::MC>()) |
+                       ex::bulk(nperms, std::move(permute_fn)));
```
[the remark this answers is not in the document]
Comment (msimberg) [reply include/dlaf/permutations/general/impl.h:398]: The number might be large, but not larger than `num_threads`. That may of course still be too much, but just keep in mind that the `thread_pool_scheduler` specialization of bulk will not blindly create `nperms` tasks.

**Review by msimberg — APPROVED** [review]
Can't comment on algorithmic changes. Looks good otherwise.

**Comment by rasolca** [comment]
cscs-ci run

**Pull request merged**